### PR TITLE
Connection Management

### PR DESCRIPTION
**Pull request by Shubhangi-cs**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+      throw new IOException(e);
+    }
+    requestBuilder.setAuthHeader(accessToken);
+    requestBuilder.setAcceptHeader(MediaType.APPLICATION_JSON);
+    requestBuilder.setContentTypeHeader(MediaType.APPLICATION_JSON);
+    RestAPIResponse apiResponse = null;
+    apiResponse = serviceNowTableAPIClient.executeGet(requestBuilder.build());
+    if (!apiResponse.isSuccess()) {
+      LOG.error("Error - {}", getErrorMessage(apiResponse.getResponseBody()));
+      throw new IOException();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: pass the error msg as arg to IOException()

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    } else {
+      String response = null;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:135]: shouldn't we throw an exception if the API response indicates an error?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    Map<String, String> properties = new HashMap<>();
+    properties.put(io.cdap.plugin.common.ConfigUtil.NAME_USE_CONNECTION, "true");
+    properties.put(ConfigUtil.NAME_CONNECTION, connectorSpecRequest.getConnectionWithMacro());
+    String tableName = connectorSpecRequest.getPath();
+    if (tableName != null) {
+      properties.put(ServiceNowConstants.PROPERTY_TABLE_NAME, tableName);
+    }
+    specBuilder.setSchema(getSchema(tableName));
+    return specBuilder.addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE,
+                                                       properties)).
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: nit: move '.' to next line

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+  }
+
+  private Schema getSchema(String tableName) {
+    SourceQueryMode mode = SourceQueryMode.TABLE;
+    SourceValueType valueType = SourceValueType.SHOW_DISPLAY_VALUE;
+    List<ServiceNowTableInfo> tableInfo = ServiceNowInputFormat.fetchTableInfo(mode, config, tableName,
+                                                                               null, valueType,
+                                                                               null, null);
+    Schema schema = tableInfo.stream().findFirst().isPresent() ? tableInfo.stream().findFirst().get().getSchema() :
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:225]: if schema can be null, we should annonate method as Nullable, and handle null schema in the caller

File: src/main/java/io/cdap/plugin/servicenow/sink/ServiceNowSinkConfig.java [hunk 5]
```diff
@@ -247,5 +247,5 @@ private Schema convertToServiceNowCompatibleDataTypes(Schema providedFieldSchema
 
     return providedFieldSchema;
   }
-
+
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/sink/ServiceNowSinkConfig.java]: remove unnecessary whitespace that was added

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
+package io.cdap.plugin.servicenow.connector;
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java:16]: add copyright, license

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 1 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 2 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:135]: I don't think so as if apiResponse is not success it will throw an error above only.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 3 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: Moved

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 4 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:225]: Done

File: src/main/java/io/cdap/plugin/servicenow/sink/ServiceNowSinkConfig.java (hunk 5 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/sink/ServiceNowSinkConfig.java]: removed

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java (hunk 6 above)
Comment (Shubhangi-cs) [reply src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java:16]: added

File: pom.xml [hunk 7]
```diff
@@ -20,7 +20,7 @@
 
   <groupId>io.cdap.plugin</groupId>
   <artifactId>servicenow-plugins</artifactId>
-  <version>1.2.0-SNAPSHOT</version>
+  <version>1.4.3-SNAPSHOT</version>
```
Comment (rmstar) [inline pom.xml]: why was the version bumped?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    if (tableName != null) {
+      properties.put(ServiceNowConstants.PROPERTY_TABLE_NAME, tableName);
+      properties.put(Constants.Reference.REFERENCE_NAME, ReferenceNames.cleanseReferenceName(tableName));
+    }
+    Schema schema = getSchema(tableName);
+    if (schema != null) {
+      specBuilder.setSchema(schema);
+    }
+    return specBuilder.addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE,
+        properties))
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: nit: fix indent

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java [hunk 9]
```diff
@@ -63,6 +68,7 @@ public ServiceNowSource(ServiceNowSourceConfig conf) {
     this.conf = conf;
   }
 
+
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java]: nit: remove extra newline

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java [hunk 10]
```diff
@@ -57,8 +56,6 @@ public class ServiceNowSourceConfig extends ServiceNowBaseSourceConfig {
   @Description("The name of the ServiceNow table from which data to be fetched. Note, the Table name value " +
     "will be ignored if the Mode is set to `Reporting`.")
   private String tableName;
-
-
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java]: nit: revert whitespace changes in this file

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        .addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSink.PLUGIN_TYPE, properties)).build();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: should we also add multi source here?

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    Set<PluginSpec> relatedPlugins = connectorSpec.getRelatedPlugins();
+    Assert.assertEquals(2, relatedPlugins.size());
+    PluginSpec pluginSpec = relatedPlugins.iterator().next();
+    Assert.assertEquals(ServiceNowConstants.PLUGIN_NAME, pluginSpec.getName());
+    Assert.assertEquals(BatchSource.PLUGIN_TYPE, pluginSpec.getType());
+    Map<String, String> properties = pluginSpec.getProperties();
+    Assert.assertEquals("true", properties.get(ConfigUtil.NAME_USE_CONNECTION));
+    Assert.assertEquals("${conn(connection-id)}", properties.get(ConfigUtil.NAME_CONNECTION));
+    Assert.assertEquals(0, collector.getValidationFailures().size());
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java]: collector is not used!

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    list.add(serviceNowTableInfo);
+    PowerMockito.mockStatic(ServiceNowInputFormat.class);
+    SourceQueryMode mode = SourceQueryMode.TABLE;
+    SourceValueType valueType = SourceValueType.SHOW_DISPLAY_VALUE;
+    Mockito.when(ServiceNowInputFormat.fetchTableInfo(mode, serviceNowSourceConfig.getConnection(),
+                                                      serviceNowSourceConfig.getTableName(),
+                                                      null, valueType,
+                                                      null, null)).thenReturn(list);
+
+    ConnectorSpec connectorSpec = serviceNowConnector.generateSpec(new MockConnectorContext
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java:169]: We should have separate test cases for each method that's being tested.
Split each test case into 3 sections:
- Test setup (create mocks, objects required to execute the test case)
- Method that you are testing. e.g: serviceNowConnector.generateSpec()
- Validate that the test did what we expected. Assert that methods were called, failures added in collector etc.

File: pom.xml (hunk 7 above)
Comment (Shubhangi-cs) [reply pom.xml]: changed back.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 8 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java (hunk 9 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSource.java]: removed

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java (hunk 10 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java]: done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 11 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: we are not supporting browse functionality in multisource.

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java (hunk 12 above)
Comment (Shubhangi-cs) [reply src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java]: removed

File: src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java (hunk 13 above)
Comment (Shubhangi-cs) [reply src/test/java/io/cdap/plugin/servicenow/connector/ServiceNowConnectorTest.java:169]: done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(ServiceNowConstants.PLUGIN_NAME)
+@Description("Connection to access data in Servicenow tables.")
+public class ServiceNowConnector implements DirectConnector {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowConnector.class);
+  private static final String OBJECT_TABLE_LIST = "sys_db_object";
+  private static final String ENTITY_TYPE_TABLE = "table";
+  private static final String LABEL_NAME = "label";
+  private static final Gson gson = new Gson();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: gson -> GSON

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 11 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: I see, so we support connections but not browse functionality for multisource - is that correct?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 8 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: still don't see this fixed

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+      }
+    }
+    return recordList;
+
+  }
+
+  @Nullable
+  private Schema getSchema(String tableName) {
+    SourceQueryMode mode = SourceQueryMode.TABLE;
+    SourceValueType valueType = SourceValueType.SHOW_DISPLAY_VALUE;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:221]: Are we using `SHOW_ACTUAL_VALUE` anywhere? https://github.com/data-integrations/servicenow-plugins/blob/develop/src/main/java/io/cdap/plugin/servicenow/util/SourceValueType.java#L32

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 14 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: done

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 8 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: Pls check, it's fixed.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 11 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: yes browse is not supported in multiSource as it does not allow user to select multi tables from the list.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 15 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:221]: No, I can't see any usage in the code.

Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:221]: Should we remove it if it's unused?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 8 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: I meant this part:
```
    return specBuilder.addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE,
        properties))
```

**Comment by rmstar** [comment]
Please make sure e2e tests are passing

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 8 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:156]: Done, looks fine now?

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java (hunk 15 above)
Comment (Shubhangi-cs) [reply src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java:221]: We should not, as Servicenow provides two types of Source Values. So it's better to always differentiate bw them that which type we are using.

File: src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+    if (tableName != null) {
+      properties.put(ServiceNowConstants.PROPERTY_TABLE_NAME, tableName);
+      properties.put(Constants.Reference.REFERENCE_NAME, ReferenceNames.cleanseReferenceName(tableName));
+    }
+    Schema schema = getSchema(tableName);
+    if (schema != null) {
+      specBuilder.setSchema(schema);
+    }
+    return specBuilder.addRelatedPlugin(new PluginSpec(ServiceNowConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE,
+                                                                                        properties))
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/connector/ServiceNowConnector.java]: nit: indent

**Review by rmstar — APPROVED** [review]
please squash commit after addressing minor comment
